### PR TITLE
Geocoder

### PR DESCRIPTION
**Pull request by jmarin**
Adds cascading geocoder to parser, addresspoints and census APIs. Provides public API (single /geocode endpoint for now). Updates and fixes several small issues with underlying services and tests

### CODE REVIEW

File: census/src/main/scala/grasshopper/census/search/AddressInterpolator.scala [hunk 1]
```diff
@@ -8,9 +8,37 @@ import grasshopper.census.model.AddressRange
...
+    }
+    val leftFromIsEven = lFromEven != "" && leftFromRange % 2 == 0
+
+    val ltoEven = f.values.getOrElse("LTOHN", 0)
+    val leftToRange = ltoEven match {
+      case "" => 0
+      case _ => ltoEven.toString.toInt
+    }
+    val leftToIsEven = ltoEven != "" && leftToRange % 2 == 0
+    val leftRangeIsEven = leftFromIsEven || leftToIsEven
```
Comment (hkeeler) [inline census/src/main/scala/grasshopper/census/search/AddressInterpolator.scala]: Looks like near-identical code repeated 4x.  Seems like a separate function is in order.

File: census/src/main/scala/grasshopper/census/search/CensusGeocode.scala [hunk 2]
```diff
@@ -34,6 +33,86 @@ trait CensusGeocode {
     }
   }
 
+  // An alternative geocoding implementation, uses less Elasticsearch query and more filtering in code. Not using for now
```
Comment (hkeeler) [inline census/src/main/scala/grasshopper/census/search/CensusGeocode.scala]: Do you think we'll end up using this function at some point?  If not, move it over to a dev branch somewhere?  If yes, can we rename the 2 `geocodeLine` functions to better describe what each does, and a scaladoc comment as to why you'd use one vs. the other?

File: census/src/main/scala/grasshopper/census/search/CensusGeocode.scala [hunk 3]
```diff
@@ -34,6 +33,86 @@ trait CensusGeocode {
...
+        .map { s =>
+          val line = s.parseJson.convertTo[Feature]
+          line
+        }
+      //lines.foreach(l => println(l.get("ZIPL")))
+      lines
+        .filter(l => zipFilter(zipCode.toString, l))
+        .filter(l => rangeFilter(number, l))
+        .map { line =>
+          println(line)
```
Comment (hkeeler) [inline census/src/main/scala/grasshopper/census/search/CensusGeocode.scala]: Looks like there's a debugging `println()` left in here.

File: census/src/main/scala/grasshopper/census/search/CensusGeocode.scala [hunk 4]
```diff
@@ -34,6 +33,86 @@ trait CensusGeocode {
...
+          val addressRange = AddressInterpolator.calculateAddressRange(line, number)
+          println(addressRange)
```
Comment (hkeeler) [inline census/src/main/scala/grasshopper/census/search/CensusGeocode.scala]: ...and here.

File: client/src/main/scala/grasshopper/client/ServiceClient.scala [hunk 5]
```diff
@@ -3,26 +3,46 @@ package grasshopper.client
...
+          ServiceUnavailable,
+          Nil,
+          HttpEntity.empty(ContentTypes.NoContentType),
+          HttpProtocols.`HTTP/1.1`
+        )
+    }
+  }
+
+  def sendResponseError(response: HttpResponse) = {
+    Future.successful {
```
Comment (hkeeler) [inline client/src/main/scala/grasshopper/client/ServiceClient.scala:43]: Should we log the error response here, or is that already taken care of elsewhere by Akka?  Before you were including `${entity}` in the error message.  Probably useful to log that as well?  Just want to make sure we know when things are behaving badly.

File: project/GrasshopperBuild.scala [hunk 6]
```diff
@@ -18,7 +18,7 @@ object BuildSettings {
       organization  := buildOrganization,
       version       := buildVersion,
       scalaVersion  := buildScalaVersion,
-      wartremoverWarnings ++= Warts.allBut(Wart.NoNeedForMonad, Wart.NonUnitStatements),
+      //wartremoverWarnings ++= Warts.allBut(Wart.NoNeedForMonad, Wart.NonUnitStatements),
```
Comment (hkeeler) [inline project/GrasshopperBuild.scala:21]: Removing the wart remover doesn't actually remove the warts. :)  Did you mean to leave this commented out?

**Comment by hkeeler** [comment]
Code review complete.  Looks good.  Just a couple open questions.

The response format is a little different than I expected, but I think in makes sense for this phase of the project where we probably _do_ want the lower-level details about the underlying services to come through.  In the future, as things stabilize and we add more data sources and services, we should consider something more generic so clients of this API don't have to know/care about `addressPointsService`, `censusService`, `xyzGeoService`.

File: project/GrasshopperBuild.scala (hunk 6 above)
Comment (jmarin) [inline project/GrasshopperBuild.scala:21]: It's producing what I consider a lot of false positives, creating a lot of noise. I want to revisit this and dig deeper, but yes, I commented out on purpose

File: client/src/main/scala/grasshopper/client/ServiceClient.scala (hunk 5 above)
Comment (jmarin) [inline client/src/main/scala/grasshopper/client/ServiceClient.scala:43]: Logging is something we need to emphasize in a subsequent milestone. I already found a couple of things that could be improved, that are related to running the project in containers. The current logstash-logback integration is working OK, but not great. When we tackle this one I would do a pass on all the code and revisit logging. Creating ticket for this

Comment (hkeeler) [inline client/src/main/scala/grasshopper/client/ServiceClient.scala:43]: Sounds good.

File: project/GrasshopperBuild.scala (hunk 6 above)
Comment (hkeeler) [inline project/GrasshopperBuild.scala:21]: OK.

**Comment by hkeeler** [comment]
I think we're good.  Merging...

**Pull request merged**